### DISCUSSION TITLE
New router utility functions for picking the first or all query parameter values

### DISCUSSION THREAD

**Discussion by mohsen1** [vercel/next.js#40151] [Ideas]
### TLDR
At a very high level:

```ts
// For URL: example.com/pizza?crust=thin&topping=pepperoni&topping=sausage 
const curst = useRouter().query.crust as string;
const toppings = useRouter().query.topping as string[];

// Add a new API to allow:
const crust = useRouter().getQuery("crust"); // type string | undefined
const toppings = useRouter().getQueryAll("topping"); // type string[] | undefined
```
### Details
I see a lot of times Next.js users mistakingly type-cast the value of `router.query.searchQueryName` to satisfy TypeScript:

```ts
const id = useRouter().query.id as string;
```

`ParsedUrlQuery` is correctly typed to have values that are either `string` or `string[]`. Because a URL like `https://example.com?searchQueryName=123&searchQueryName=456` is valid and if you using the standard URL APIs you can get both values:

```ts
const url = new URL("https://example.com?searchQueryName=123&searchQueryName=456");
url.searchParams.get("searchQueryName"); //  "123"
url.searchParams.getAll("searchQueryName"); // ["123", "456"]
```

The `URLSearchParams` API returns the first query param when calling the `get()` method. 

### `pickFirst`
In our codebase we have a function called `pickFirst` to go around this type-cast situation:

```ts
function pickFirst(queryParam: string | string[] | undefined) {
  if (Array.isArray(queryParam)) {
    return queryParam[0];
  }
  return queryParam;
}
```

```ts
// const id = useRouter().query.id as string;
const id = pickFirst(useRouter().query.id); // type string | undefined
```

## Suggestion
put the `pickFirst` logic into the router:

```ts
const id = useRouter().getQuery("id"); // type string | undefined
const toppings = useRouter().getQueryAll("topping"); // type string[] | undefined
```

### Goals
* use the same terminology as `URLSearchParams`
* do not break existing API 
* recommend using this new API in the docs instead of `router.query`

**Comment by JMartinCollins** [comment]
For any query param that could be an array, it could also just as easily have a single value:
`https://example.com?selectedId=1`
`https://example.com?selectedId=1&selectedId=2`
And this requires the developer to implement logic for both the `string` and `string[]` cases, which is more work and just asking for bugs. The solution put forth here is definitely moving in the right direction. It's certainly too late to implement now, but it would be more convenient for the param in `router.query.param` to always be a `string[]` or `undefined`, and never `string`.
Perhaps a solution might be to allow for some kind of middleware or configuration option to customize the parsing of query params in `next.config.js`. That would allow changing the default behavior of returning `string | string[]` to just return `string[]`.
